### PR TITLE
Break intermediate measurements on 3+ qubits into single qubit measurements in `RouteCQC`#6293

### PR DESCRIPTION
**Pull request by shef4**
### router mishandling mid-circuit measurements #issue6293
**Description of the issue**

`cirq.RouteCQC` commutes gates in undefined ways due to intermediate measurements with 3 or more qubits not being handled correctly.

**This PR aims to fix this issue by:**

Breaking intermediate measurements on 3+ qubits into single qubit measurements if the result isn't stored (i.e. key is None) else a `ValueError` is raised. 

**Next Steps**

Check if it's possible to create a [ClassicallyControlledOperation](https://quantumai.google/reference/python/cirq/ClassicallyControlledOperation) controlled by the result of a multiqubit measurement
```python
cirq.ClassicallyControlledOperation(
    sub_operation: 'cirq.Operation',
    conditions: Sequence[Union[str, 'cirq.MeasurementKey', 'cirq.Condition', sympy.Basic]]
)
``` 
From looking at the documentation it seems to be possible. Will stop here until there is a need to support this case.

### CODE REVIEW

File: cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py [hunk 1]
```diff
@@ -107,6 +107,53 @@ def test_circuit_with_measurement_gates():
...
+
+
+def test_circuit_with_multi_qubit_intermediate_measurement_gate_and_result_not_stored():
+    device = cirq.testing.construct_ring_device(3)
+    device_graph = device.metadata.nx_graph
+    router = cirq.RouteCQC(device_graph)
+    q = cirq.LineQubit.range(3)
+    hard_coded_mapper = cirq.HardCodedInitialMapper({q[i]: q[i] for i in range(3)})
+    circuit = cirq.Circuit(
+        [cirq.Moment(cirq.MeasurementGate(3).on(*q)), cirq.Moment(cirq.H.on_each(q))]
```
Comment (NoureldinYosri) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py:130]: don't construct measurement gates manually. use `cirq.measure` instead

File: cirq-core/cirq/transformers/routing/route_circuit_cqc.py [hunk 2]
```diff
@@ -245,17 +245,32 @@ def _get_one_and_two_qubit_ops_as_timesteps(
...
         output routed circuit, single-qubit operations are inserted before two-qubit operations.
+
+        Raises:
+            ValueError: if circuit has intermediate measurement op's that act on three or more qubits and result is stored.
         """
         two_qubit_circuit = circuits.Circuit()
         single_qubit_ops: List[List[cirq.Operation]] = []
+        current_moment = 0
+
         for moment in circuit:
```
Comment (NoureldinYosri) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc.py:256]:
```py3
for i, moment in enumerate(circuit):
```

File: cirq-core/cirq/transformers/routing/route_circuit_cqc.py [hunk 3]
```diff
@@ -245,17 +245,32 @@ def _get_one_and_two_qubit_ops_as_timesteps(
...
+            current_moment += 1
             for op in moment:
                 timestep = two_qubit_circuit.earliest_available_moment(op)
                 single_qubit_ops.extend([] for _ in range(timestep + 1 - len(single_qubit_ops)))
                 two_qubit_circuit.append(
                     circuits.Moment() for _ in range(timestep + 1 - len(two_qubit_circuit))
                 )
-                if protocols.num_qubits(op) == 2 and not protocols.is_measurement(op):
+                if protocols.num_qubits(op) > 2 and isinstance(op.gate, ops.MeasurementGate):
```
Comment (NoureldinYosri) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc.py:264]: keep `protocols.is_measurement`

File: cirq-core/cirq/transformers/routing/route_circuit_cqc.py [hunk 4]
```diff
@@ -245,17 +245,32 @@ def _get_one_and_two_qubit_ops_as_timesteps(
...
+                    if len(circuit.moments) == current_moment:
+                        single_qubit_ops[timestep].append(op)
+                    elif op.gate.key == "":
+                        single_qubit_ops[timestep] += [ops.measure(qubit) for qubit in op.qubits]
```
Comment (NoureldinYosri) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc.py:268]:
```py3
single_qubit_ops[timestep].extend(ops.measure(qubit) for qubit in op.qubits)
```

File: cirq-core/cirq/transformers/routing/route_circuit_cqc.py [hunk 5]
```diff
@@ -245,17 +245,32 @@ def _get_one_and_two_qubit_ops_as_timesteps(
...
                 timestep = two_qubit_circuit.earliest_available_moment(op)
                 single_qubit_ops.extend([] for _ in range(timestep + 1 - len(single_qubit_ops)))
                 two_qubit_circuit.append(
                     circuits.Moment() for _ in range(timestep + 1 - len(two_qubit_circuit))
                 )
-                if protocols.num_qubits(op) == 2 and not protocols.is_measurement(op):
+                if protocols.num_qubits(op) > 2 and isinstance(op.gate, ops.MeasurementGate):
+                    if len(circuit.moments) == current_moment:
+                        single_qubit_ops[timestep].append(op)
+                    elif op.gate.key == "":
```
Comment (NoureldinYosri) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc.py:267]: the default for `key` is not an empty string it's `','.join(str(q) for q in qubits)`. instead make it
```py3
elif op.gate.key in ('', ops.measure(op.qubits).gate.key):
```

File: docs/transform/routing_transformer.ipynb [hunk 6]
```diff
@@ -465,6 +465,18 @@
   "kernelspec": {
    "display_name": "Python 3",
    "name": "python3"
+  },
+  "language_info": {
```
Comment (shef4) [inline docs/transform/routing_transformer.ipynb:469]: need to remove

**Review by NoureldinYosri — CHANGES_REQUESTED** [review]
please remove changes to unrelated files

File: cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py [hunk 7]
```diff
@@ -107,6 +107,56 @@ def test_circuit_with_measurement_gates():
...
+
+def test_circuit_with_multi_qubit_intermediate_measurement_gate_and_result_not_stored():
+    device = cirq.testing.construct_ring_device(3)
+    device_graph = device.metadata.nx_graph
+    router = cirq.RouteCQC(device_graph)
+    q = cirq.LineQubit.range(3)
+    hard_coded_mapper = cirq.HardCodedInitialMapper({q[i]: q[i] for i in range(3)})
+    circuit = cirq.Circuit(
+        [
+            cirq.Moment(cirq.measure(q[0]), cirq.measure(q[1]), cirq.measure(q[2])),
```
Comment (NoureldinYosri) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py:131]: these are three measurements

File: cirq-core/cirq/transformers/routing/route_circuit_cqc.py [hunk 8]
```diff
@@ -245,17 +245,32 @@ def _get_one_and_two_qubit_ops_as_timesteps(
...
         gates and single-qubit gates of the i'th timesteps respectively. When constructing the
         output routed circuit, single-qubit operations are inserted before two-qubit operations.
+
+        Raises:
+            ValueError: if circuit has intermediate measurement op's that act on three or more qubits and result is stored.
         """
         two_qubit_circuit = circuits.Circuit()
         single_qubit_ops: List[List[cirq.Operation]] = []
-        for moment in circuit:
+        current_moment = 0
```
Comment (NoureldinYosri) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc.py:254]: this is not needed

File: cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py [hunk 9]
```diff
@@ -107,6 +107,56 @@ def test_circuit_with_measurement_gates():
     cirq.testing.assert_same_circuits(routed_circuit, circuit)
 
 
+def test_circuit_with_two_qubit_intermediate_measurement_gate():
+    device = cirq.testing.construct_ring_device(2)
+    device_graph = device.metadata.nx_graph
+    router = cirq.RouteCQC(device_graph)
+    q = cirq.LineQubit.range(2)
+    hard_coded_mapper = cirq.HardCodedInitialMapper({q[i]: q[i] for i in range(2)})
+    circuit = cirq.Circuit(cirq.measure_each(*q), cirq.H.on_each(q))
```
Comment (NoureldinYosri) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py:116]: `cirq.measure_each(qs)` is different from `cirq.measure(qs)`

File: cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py [hunk 10]
```diff
@@ -107,6 +107,56 @@ def test_circuit_with_measurement_gates():
...
+    cirq.testing.assert_same_circuits(routed_circuit, expected)
+
+
+def test_circuit_with_multi_qubit_intermediate_measurement_gate_and_result_stored():
+    device = cirq.testing.construct_ring_device(3)
+    device_graph = device.metadata.nx_graph
+    router = cirq.RouteCQC(device_graph)
+    q = cirq.LineQubit.range(3)
+    hard_coded_mapper = cirq.HardCodedInitialMapper({q[i]: q[i] for i in range(3)})
+    circuit = cirq.Circuit(cirq.MeasurementGate(3, key="key").on(*q), cirq.H.on_each(*q))
```
Comment (NoureldinYosri) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py:153]: cirq.measure(qs)

File: cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py [hunk 11]
```diff
@@ -107,6 +107,56 @@ def test_circuit_with_measurement_gates():
...
+        ]
+    )
+    cirq.testing.assert_same_circuits(routed_circuit, expected)
+
+
+def test_circuit_with_multi_qubit_intermediate_measurement_gate_and_result_stored():
+    device = cirq.testing.construct_ring_device(3)
+    device_graph = device.metadata.nx_graph
+    router = cirq.RouteCQC(device_graph)
+    q = cirq.LineQubit.range(3)
```
Comment (NoureldinYosri) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py:151]: `qs`

File: docs/transform/routing_transformer.ipynb [hunk 12]
```diff
@@ -75,7 +75,7 @@
   },
   {
    "cell_type": "code",
-   "execution_count": null,
+   "execution_count": 1,
```
Comment (NoureldinYosri) [inline docs/transform/routing_transformer.ipynb:78]: please remove your changes to this file

**Review by NoureldinYosri — CHANGES_REQUESTED** [review]

File: cirq-core/cirq/transformers/routing/route_circuit_cqc.py (hunk 3 above)
Comment (shef4) [reply cirq-core/cirq/transformers/routing/route_circuit_cqc.py]: gett back error this from ./check/mypy when I use protocols.is_measurement

cirq-core/cirq/transformers/routing/route_circuit_cqc.py:265: error: Item "Gate" of "Optional[Gate]" has no attribute "key"  [union-attr]
cirq-core/cirq/transformers/routing/route_circuit_cqc.py:265: error: Item "None" of "Optional[Gate]" has no attribute "key"  [union-attr]

File: cirq-core/cirq/transformers/routing/route_circuit_cqc.py [hunk 13]
```diff
@@ -245,17 +245,30 @@ def _get_one_and_two_qubit_ops_as_timesteps(
...
                 timestep = two_qubit_circuit.earliest_available_moment(op)
                 single_qubit_ops.extend([] for _ in range(timestep + 1 - len(single_qubit_ops)))
                 two_qubit_circuit.append(
                     circuits.Moment() for _ in range(timestep + 1 - len(two_qubit_circuit))
                 )
-                if protocols.num_qubits(op) == 2 and not protocols.is_measurement(op):
+                if protocols.num_qubits(op) > 2 and protocols.is_measurement(op):
+                    if len(circuit.moments) == i + 1:
+                        single_qubit_ops[timestep].append(op)
+                    elif op.gate.key in ('', ops.measure(op.qubits).gate.key):
```
Comment (shef4) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc.py:265]: getting back error this from ./check/mypy when I use protocols.is_measurement


```python
cirq-core/cirq/transformers/routing/route_circuit_cqc.py:265: error: Item "Gate" of "Optional[Gate]" has no attribute "key"  [union-attr]
cirq-core/cirq/transformers/routing/route_circuit_cqc.py:265: error: Item "None" of "Optional[Gate]" has no attribute "key"  [union-attr]
```

Comment (NoureldinYosri) [reply cirq-core/cirq/transformers/routing/route_circuit_cqc.py:265]: create 
```py3
key = op.gate.key  # type: ignore
```
then use it in your code

File: cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py [hunk 14]
```diff
@@ -107,6 +107,48 @@ def test_circuit_with_measurement_gates():
...
+
+
+def test_circuit_with_multi_qubit_intermediate_measurement_gate_and_result_stored():
+    device = cirq.testing.construct_ring_device(3)
+    device_graph = device.metadata.nx_graph
+    router = cirq.RouteCQC(device_graph)
+    qs = cirq.LineQubit.range(3)
+    hard_coded_mapper = cirq.HardCodedInitialMapper({qs[i]: qs[i] for i in range(3)})
+    circuit = cirq.Circuit(
+        [cirq.Moment(cirq.measure(qs, key="key_name")), cirq.Moment(cirq.H.on_each(qs))]
```
Comment (NoureldinYosri) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py:144]: nit: `key='test'`

File: cirq-core/cirq/transformers/routing/route_circuit_cqc.py [hunk 15]
```diff
@@ -245,17 +245,32 @@ def _get_one_and_two_qubit_ops_as_timesteps(
         The i'th entry in the nested two-qubit and single-qubit ops correspond to the two-qubit
         gates and single-qubit gates of the i'th timesteps respectively. When constructing the
         output routed circuit, single-qubit operations are inserted before two-qubit operations.
+
+        Raises:
+            ValueError: if circuit has intermediate measurement op's that act on three or more qubits and result is stored.
```
Comment (NoureldinYosri) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc.py:250]: `ValueError: if circuit has intermediate measurements that act on three or more qubits with a custom key.`

File: cirq-core/cirq/transformers/routing/route_circuit_cqc.py [hunk 16]
```diff
@@ -245,17 +245,32 @@ def _get_one_and_two_qubit_ops_as_timesteps(
...
+                if protocols.num_qubits(op) > 2 and protocols.is_measurement(op):
+                    key = op.gate.key  # type: ignore
+                    default_key = ops.measure(op.qubits).gate.key  # type: ignore
+                    if len(circuit.moments) == i + 1:
+                        single_qubit_ops[timestep].append(op)
+                    elif key in ('', default_key):
+                        single_qubit_ops[timestep].extend(ops.measure(qubit) for qubit in op.qubits)
+                    else:
+                        raise ValueError(
+                            'Non-terminal measurements on three or more qubits when result is stored are not supported'
```
Comment (NoureldinYosri) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc.py:271]: `'Intermediate measurements on three or more qubits with a custom key are not supported'`

**Review by NoureldinYosri — CHANGES_REQUESTED** [review]
just a few more nits + please fix the conflict with the main branch

**Review by NoureldinYosri — CHANGES_REQUESTED** [review]
almost there. one small change

**Review by NoureldinYosri — APPROVED** [review]

**Pull request merged**